kamon-core/src/main/scala/kamon/context/HttpPropagation.scala

```diff
@@ -0,0 +1,283 @@
+  }
+
+  /**
+    * Encapsulates logic required to write a single context entry to HTTP headers. Implementations of this trait
+    * must be aware of the entry they are able to write and the HTTP headers required to do so.
+    */
+  trait EntryWriter {
+
+    /**
+      * Tries to write a context entry into HTTP headers.
+      *
+      * @param context The context from which entries should be written.
+      * @param writer Wrapper on the HTTP message that will carry the context headers.
+      * @param direction Write direction. It can be either Outgoing or Returning.
+      */
+    def writeEntry(context: Context, writer: HttpPropagation.HeaderWriter, direction: Direction.Write): Unit
+  }
+
+
+  /**
+    * Wrapper that reads HTTP headers from HTTP a message.
+    */
+  trait HeaderReader {
+
+    /**
+      * Reads an HTTP header value
+      *
+      * @param header HTTP header name
+      * @return The HTTP header value, if present.
+      */
+    def readHeader(header: String): Option[String]
+  }
+
+  /**
+    * Wrapper that writes HTTP headers to a HTTP message.
+    */
+  trait HeaderWriter {
+
+    /**
+      * Writes a HTTP header into a HTTP message.
+      *
+      * @param header HTTP header name.
+      * @param value HTTP header value.
+      */
+    def writeHeader(header: String, value: String): Unit
+  }
+
+
+  /**
+    * Create a new default HttpPropagation instance from the provided configuration.
+    *
+    * @param config HTTP propagation channel configuration
+    * @return A newly constructed HttpPropagation instance.
+    */
+  def from(config: Config, classLoading: ClassLoading): HttpPropagation = {
+    new HttpPropagation.Default(Components.from(config, classLoading))
+  }
+
+  /**
+    * Default HTTP Propagation in Kamon.
+    */
+  final class Default(components: Components) extends HttpPropagation {
+    private val log = LoggerFactory.getLogger(classOf[HttpPropagation.Default])
+
+    /**
+      * Reads context tags and entries on the following order:
+      *   - Read all context tags from the context tags header.
+      *   - Read all context tags with explicit mappings. This overrides any tag from the previous step in case
+      *     of a tag key clash.
+      *   - Read all context entries using the incoming entries configuration.
+      */
+    override def readContext(reader: HeaderReader): Context = {
+      val tags = Map.newBuilder[String, String]
+
+      // Tags encoded together in the context tags header.
+      try {
+        reader.readHeader(components.tagsHeaderName).foreach { contextTagsHeader =>
+          contextTagsHeader.split(";").foreach(tagData => {
+            val tagPair = tagData.split("=")
+            if (tagPair.length == 2) {
+              tags += (tagPair(0) -> tagPair(1))
+            }
+          })
+        }
+      } catch {
+        case t: Throwable => log.warn("Failed to read the context tags header", t.asInstanceOf[Any])
```

Comment:
maybe we can use `NonFatal(t)` instead of catch `Throwable`?

Reply:
done!